src/utils/dockerDirManager.ts

```diff
@@ -181,12 +182,58 @@ export default class DockerDirManager {
     return this.contextName;
   }
 
+  /**
+   * Determines whether the passed credential helper is working.
+   * @param helperName The cred helper name, without the "docker-credential-" prefix.
+   */
+  protected async credHelperWorking(helperName: string): Promise<boolean> {
+    const helperBin = `docker-credential-${ helperName }`;
+    const platform = os.platform();
+    let pathVar = process.env.PATH ?? '';
+
+    if (helperName === 'desktop') {
+      // Special case docker-credentials-desktop: never use it.
+      return false;
+    }
+
+    // The PATH needs to contain our resources directory (on macOS that would
+    // not be in the application's PATH), as well as /usr/local/bin.
+    // NOTE: This needs to match HttpCredentialHelperServer.
+    pathVar += path.delimiter + path.join(paths.resources, platform, 'bin');
+    if (platform === 'darwin') {
+      pathVar += `${ path.delimiter }/usr/local/bin`;
+    }
+
+    try {
+      // Provide input in case the helper always reads from stdin regardless of argument (harmless if it doesn't).
+      const body = stream.Readable.from('');
+
+      await spawnFile(helperBin, ['list'], {
+        env:   { ...process.env, PATH: pathVar },
+        stdio: [body, 'pipe', console],
+      });
+
+      return true;
+    } catch (err) {
+      console.log(`Credential helper "${ helperBin }" is not functional: ${ err }`);
+
+      return false;
+    }
+  }
+
   /**
    * Returns the default cred helper name for the current platform.
    */
-  protected getCredsStoreFor(currentCredsStore: string | undefined): string {
+  protected async getCredsStoreFor(currentCredsStore: string | undefined): Promise<string> {
     const platform = os.platform();
 
+    // If the custom credential helper exists, use it.  Note that this may
+    // sometimes fail if the user is in a shell with a different PATH than our
+    // process, but we can't help with that right now.
+    if (currentCredsStore && await this.credHelperWorking(currentCredsStore)) {
```

Comment:
I see later in this function this test:

```
      if (currentCredsStore === 'secretservice') {
        return 'secretservice';
      } else {
        return 'pass';
      }
```

Why would we do this, given that we have checked here that the helper doesn't actually work?

But assuming we want to do this for some reason, why does the later test not need to also check if `currentCredsStore` is defined first as well?